Implement DecomposeRotate

adiaaida opened the pull request:
fgRecognizeAndMorphBitwiseRotation can potentially morph a tree with a
TYP_LONG return type on x86 if the rotate amount is a GT_CNS_INT. This
change implements DecomposeRotate for those rotate instructions. There are
5 cases:

1) Rotate by 0: do nothing.
2) Rotate by 32: swap hi and lo.
3) Rotate by < 32: produce a shld/shld (for GT_ROL) or a shrd/shrd (for
GT_ROR).
4) Rotate by >= 32: swap hi and lo, subtract from the rotate amount 32, and
do option 3.
5) Rotate by >= 64: do rotateAmount % 64 to get the rotate amount between
0 and 63, then do one of options 1-4.

adiaaida commented:
@BruceForstall @pgavlin PTAL
/cc: @dotnet/jit-contrib

mikedn commented on src/jit/decomposelongs.cpp:
```diff
@@ -1294,6 +1299,140 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+        GenTree* next = tree->gtNext;
+        // Remove tree and don't do anything else.
+        Range().Remove(tree);
+        use.ReplaceWith(m_compiler, gtLong);
```
Hmm, didn't you remove `gtLong` above?

adiaaida replied:
Good catch

mikedn commented on src/jit/decomposelongs.cpp:
```diff
@@ -1294,6 +1299,140 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+        // If the rotate amount is 32, then swap hi and lo
+        hiResult = loOp1;
+        loResult = hiOp1;
+        Range().InsertBefore(tree, loResult, hiResult);
```
This could reuse the existing `gtLong` and swap its operands.

adiaaida replied:
Done.

adiaaida commented on src/jit/decomposelongs.cpp:
```diff
@@ -1294,6 +1299,140 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+    Range().Remove(rotateByOp);
+
+    // Make sure the rotate amount is between 0 and 63.
+    count = count % 64;
```
I'm not sure. I didn't see it in my test case, but I wanted to cover my bases.

BruceForstall replied:
The patterns in fgRecognizeAndMorphBitwiseRotation() look pretty clearly to never generate count > 64 (or 32 for TYP_INT), so you could assert this. I think you can assert count !=0 as well.

---
In reply to: [84173232](https://github.com/dotnet/coreclr/pull/7724#discussion_r84173232) [](ancestors = 84173232)

BruceForstall commented:
I assume this is for VSO 278370.

It would be nice to add a standalone test case to the open source test suite.

adiaaida commented:
@BruceForstall Yes it is. I wasn't sure how to reproduce the IR pattern that led to the failure, but I agree we should check in a test case.

BruceForstall commented on src/jit/decomposelongs.cpp:
```diff
@@ -1294,6 +1299,144 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
+
+    assert((oper == GT_ROL) || (oper == GT_ROR));
+    assert((rotateByOper == GT_CNS_INT));
+
```
nit: don't need 2 sets of parens.

BruceForstall replied:
since you don't otherwise use rotateByOper, you could just: assert(rotateByOp->IsCnsIntOrI()) and delete it.

---
In reply to: [84177632](https://github.com/dotnet/coreclr/pull/7724#discussion_r84177632) [](ancestors = 84177632)

adiaaida replied:
Done.

adiaaida commented:
@dotnet-bot retest this please

adiaaida commented:
@BruceForstall It turns out we have a test that is supposed to be testing this functionality, but the way it was written, we end up folding the constants, eliminating the rol. I'm working on adding an additional case to the Rotate.cs test to actually test this functionality.

adiaaida commented:
@BruceForstall I have updated the Rotate test in CodeGenBringUpTests to test the functionality of this change. I have also updated this change to fix a bug in it that I found when running the Rotate tests.

PTAL

BruceForstall commented on tests/src/JIT/CodeGenBringUpTests/Rotate.cs:
```diff
@@ -128,12 +146,30 @@ static ulong ror64comm(ulong value, int amount)
     {
-        ulong value = flag() ? (ulong)0x123456789abcdef : (ulong)0x123456789abcdef;
+        ulong value = flag() ? (ulong)0x123456789abcdef : (ulong)0xabcdef123456789;
         int amount = flag() ? 5 : 5;
```
It seems like the line below:
```
int amount = flag() ? 5 : 5;
```
is intended to make the rotate by a variable amount? Don't we optimize that away, though, and then constant fold the shift amounts? Or, the name "ror64const" I guess implies the rotation amount should be constant, then why do we need to `?:`?

adiaaida replied:
I'm not sure. @erozenfeld wrote the initial test. May just be a copy-paste error.

adiaaida replied:
It's interesting that that isn't the case for rol. I say leave it for now.

BruceForstall commented:
LGTM
